text/0000-offset_of.md

```diff
@@ -0,0 +1,576 @@
+// which can be more readable if you need to use them several times.
+const OFFSET_VERTEX_TEX: usize = core::mem::offset_of!(Vertex, tex);
+const OFFSET_VERTEX_POS: usize = core::mem::offset_of!(Vertex, pos);
+```
+
+As you can see, the usage is the same as before, but because we didn't specify
+`#[repr(C)]`, compiler may have changed the order or position, so the values may
+be different -- it's completely possible that `pos` is located at offset 0, for
+example! Thankfully, by using `core::mem::offset_of!`, this code is correct
+either way, and will continute to be correct, even if the layout algorithm
+changes in the future.
+
+## `offset_of!` On Other Types
+
+If your type doesn't have named fields, `offset_of!` can still be used. For
+tuples and tuple structs, the "name" of the field is the numeral value you use
+to access it. For example:
+
+```rust
+// Works with a tuple struct
+struct KeyVal(&'static str, Vec<u8>);
+const OFFSET_KV_KEY: usize = core::mem::offset_of!(KeyVal, 0);
+const OFFSET_KV_VAL: usize = core::mem::offset_of!(KeyVal, 1);
+// Or with an anonymous tuple.
+const OFFSET_ANON_KEY: usize = core::mem::offset_of!((&'static str, Vec<u8>), 0);
+const OFFSET_ANON_VAL: usize = core::mem::offset_of!((&'static str, Vec<u8>), 1);
+```
+
+Finally, `offset_of!` can be used to compute the offset of fields in unions too.
+While this may be surprising, the compiler is allowed to put padding in front of
+fields in unions which are not `#[repr(C)]`, which would lead to a non-zero
+field offset.
+
+```rs
+use core::mem::offset_of;
+union Buffer {
+    metadata: [u64; 3],
+    datadata: [u8; 1024 * 1024 * 32],
+}
+const METADATA_OFFSET: usize = offset_of!(Buffer, metadata);
+```
+
+## Limitations
+
+There are a few limitations worth mentioning. Some of these may be relaxed in
+the future, however.
+
+1. Perhaps unsurprisingly, it obeys privacy, so both the type and field you call
+   `offset_of!` on must be visible to the code calling `offset_of!`.
+
+2. The type holding the field must be `Sized`, so trying to compute where the
+   slice begins in something like `offset_of!((i32, [u32]), 1)` isn't supported.
+
+3. Compared to `offsetof` in C and C++, you can't access nested fields/arrays.
+   That is, instead of `offset_of!(Foo, quank.zoop.2.quank[4])`, you'll have to
+   compute the offsets of each step manually, and sum them.
+
+4. Finally, types other than tuples, structs, and unions are currently
+   unsupported.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+`offset_of` is a new macro exported from `core::mem` which has a signature
+similar to the following:
+
+```rs
+pub macro offset_of($Container:ty, $field:tt $(,)?) {
+    // ...implementation defined...
+}
+```
+
+Invoking this macro expands to a constant expression of type `usize`, which
+evaluates to the offset in bytes from the beginning of `$Container` where
+`$field` is found.
```

Comment:
How about specifically having it return a literal? That would allow passing its result to another macro that expects a `literal` fragment, or reading its value in a proc-macro via #90765 (currently nightly-only).